sky/spot/spot_state.py

```diff
@@ -388,104 +423,122 @@ def get_nonterminal_job_ids_by_name(name: Optional[str]) -> List[int]:
+    """
+    id_statuses = _get_all_task_ids_statuses(job_id)
     if len(id_statuses) == 0:
         return None, None
-    sub_job_id, status = id_statuses[-1]
+    task_id, status = id_statuses[-1]
     status = SpotStatus(status)
-    for sub_job_id, status in id_statuses:
+    for task_id, status in id_statuses:
         status = SpotStatus(status)
         if not status.is_terminal():
             break
-    return sub_job_id, status
+    return task_id, status
 
 
 def get_status(job_id: int) -> Optional[SpotStatus]:
-    _, status = get_latest_sub_job_id_status(job_id)
+    _, status = get_latest_task_id_status(job_id)
     return status
 
 
 def get_failure_reason(job_id: int) -> Optional[str]:
-    """Get the failure reason of a job."""
+    """Get the failure reason of a job.
+
+    If the job has multiple tasks, we return the first failure reason.
+    """
     reason = _CURSOR.execute(
         """\
-        SELECT failure_reason FROM spot WHERE new_job_id=(?)""",
-        (job_id,)).fetchall()
-    reason = [r[0] for r in reason if r[0]]
+        SELECT failure_reason FROM spot
+        WHERE spot_job_id=(?)
+        ORDER BY task_id ASC""", (job_id,)).fetchall()
+    reason = [r[0] for r in reason if r[0] is not None]
     if len(reason) == 0:
         return None
     return reason[0]
 
 
 def get_spot_jobs(job_id: Optional[int] = None) -> List[Dict[str, Any]]:
     """Get spot clusters' status."""
-    job_filter = '' if job_id is None else f'WHERE spot.new_job_id={job_id}'
+    job_filter = '' if job_id is None else f'WHERE spot.spot_job_id={job_id}'
 
+    # Join the spot and job_info tables to get the job name for each task.
+    # We use LEFT OUTER JOIN mainly for backward compatibility, as for an
+    # existing controller before #1982, the job_info table may not exist,
+    # and all the spot jobs created before will not present in the
+    # job_info.
     rows = _CURSOR.execute(f"""\
         SELECT *
         FROM spot
-        LEFT OUTER JOIN job_names
-        ON spot.new_job_id=job_names.new_job_id
+        LEFT OUTER JOIN job_info
+        ON spot.spot_job_id=job_info.spot_job_id
         {job_filter}
-        ORDER BY spot.new_job_id DESC, spot.sub_job_id ASC""").fetchall()
+        ORDER BY spot.spot_job_id DESC, spot.task_id ASC""").fetchall()
     jobs = []
     for row in rows:
         job_dict = dict(zip(columns, row))
         job_dict['status'] = SpotStatus(job_dict['status'])
-        if job_dict['aggregated_job_name'] is None:
-            job_dict['aggregated_job_name'] = job_dict['job_name']
+        if job_dict['job_name'] is None:
+            job_dict['job_name'] = job_dict['task_name']
         jobs.append(job_dict)
     return jobs
 
 
-def get_sub_job_name(job_id: int, sub_job_id: int) -> str:
+def get_task_name(job_id: int, task_id: int) -> str:
     """Get the task name of a job."""
     task_name = _CURSOR.execute(
         """\
         SELECT job_name FROM spot
```

Comment:
Probably should copy the legacy note from schema def to here as well

Reply:
Ahh, good point! The `job_name` is quite confusing. I copied that column and made a new column in `spot` table, to make it more readable.